File: src/trocado/features/identity/application/use_cases/sign_up_use_case.py
from __future__ import annotations

import asyncio

from trocado.core.application.interfaces.clock_interface import ClockInterface
from trocado.core.application.interfaces.identifier_provider_interface import (
    IdentifierProviderInterface,
)
from trocado.features.identity.application.data.person_data import PersonData
from trocado.features.identity.application.data.sign_up_data import SignUpData
from trocado.features.identity.application.interfaces.password_hasher_interface import (
    PasswordHasherInterface,
)
from trocado.features.identity.application.interfaces.person_repository_interface import (
    PersonRepositoryInterface,
)
from trocado.features.identity.application.mappers.person_data_mapper import PersonDataMapper
from trocado.features.identity.domain.entities.person_entity import PersonEntity
from trocado.features.identity.domain.errors.email_already_in_use_error import EmailAlreadyInUseError
from trocado.features.identity.domain.value_objects.email_value_object import EmailValueObject
from trocado.features.identity.domain.value_objects.name_value_object import NameValueObject
from trocado.features.identity.domain.value_objects.password_value_object import PasswordValueObject


class SignUpUseCase:
    """Sign up a new person: validate input, enforce email uniqueness, hash, persist, return public data."""

    def __init__(
        self,
        clock: ClockInterface,
        hasher: PasswordHasherInterface,
        repository: PersonRepositoryInterface,
        identifier: IdentifierProviderInterface,
    ) -> None:
        self._clock = clock
        self._hasher = hasher
        self._repository = repository
        self._identifier = identifier

    async def execute(self, data: SignUpData) -> PersonData:
        name = NameValueObject(data.name)
        email = EmailValueObject(data.email)
        password = PasswordValueObject(data.password)

        if await self._repository.find_active_by_email(email) is not None:
            raise EmailAlreadyInUseError()

        created_at, identifier, password_hash = await asyncio.gather(
            self._clock.now(),
            self._identifier.generate(),
            self._hasher.hash(password),
        )

        person = PersonEntity.create(
            name=name,
            email=email,
            id=identifier,
            created_at=created_at,
            password=password_hash,
        )
        await self._repository.create(person)

        return PersonDataMapper.to_data(person)

File: src/main/kotlin/com/bed/cordato/features/identity/application/use_cases/SignUpUseCase.kt
package com.bed.cordato.features.identity.application.use_cases

import com.bed.cordato.core.application.ports.IdGeneratorPort

import com.bed.cordato.features.identity.domain.errors.SignUpError
import com.bed.cordato.features.identity.domain.entities.PersonEntity
import com.bed.cordato.features.identity.domain.enums.PersonStatusEnum

import com.bed.cordato.features.identity.domain.value_objects.EmailValueObject
import com.bed.cordato.features.identity.domain.value_objects.NameValueObject
import com.bed.cordato.features.identity.domain.value_objects.PasswordValueObject

import com.bed.cordato.features.identity.application.results.SignUpResult
import com.bed.cordato.features.identity.application.commands.SignUpCommand
import com.bed.cordato.features.identity.application.ports.PasswordHasherPort
import com.bed.cordato.features.identity.application.repositories.PersonRepository

/**
 * Registers a new person. Order is deliberate: validate each input (fail-fast,
 * e-mail first), then check e-mail uniqueness *before* hashing so a doomed signup
 * never pays the hashing cost, then hash, build the active person, and persist.
 *
 * The public `invoke` signature is the driving (primary) side of this context — no
 * extra interface is added for it.
 */
class SignUpUseCase(
    private val generator: IdGeneratorPort,
    private val hasher: PasswordHasherPort,
    private val repository: PersonRepository,
) {
    operator fun invoke(command: SignUpCommand): SignUpResult {
        val email = EmailValueObject.of(command.email)
            ?: return SignUpResult.Failure(SignUpError.InvalidEmail)

        val name = NameValueObject.of(command.name)
            ?: return SignUpResult.Failure(SignUpError.InvalidName)

        val password = PasswordValueObject.of(command.password)
            ?: return SignUpResult.Failure(SignUpError.WeakPassword(PasswordValueObject.MIN_LENGTH))

        if (repository.existsByEmail(email)) {
            return SignUpResult.Failure(SignUpError.EmailAlreadyInUse)
        }

        val person = PersonEntity(
            name = name,
            email = email,
            id = generator.invoke(),
            hash = hasher.hash(password),
            status = PersonStatusEnum.ACTIVE,
        )
        repository.save(person)

        return SignUpResult.Success(person)
    }
}
